cmd/earthly/settings_test.go

```diff
@@ -0,0 +1,354 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"testing"
+
+	"github.com/earthly/earthly/buildkitd"
+	"github.com/earthly/earthly/config"
+	"github.com/earthly/earthly/conslogging"
+	"github.com/stretchr/testify/assert"
+	"github.com/urfave/cli/v2"
+)
+
+var noopArgs = []string{""}
+
+func TestBuildArgMatrix(t *testing.T) {
+	var tests = []struct {
+		testName string
+		config   config.GlobalConfig
+		args     []string
+		expected addresses
+	}{
+		{
+			"No Config, no CLI",
+			config.GlobalConfig{
+				BuildkitHost:      "",
+				DebuggerHost:      "",
+				DebuggerPort:      config.DefaultDebuggerPort,
+				LocalRegistryHost: "",
+			},
+			noopArgs,
+			addresses{
+				buildkit:      buildkitd.DockerAddress,
+				debugger:      fmt.Sprintf("tcp://127.0.0.1:%v", config.DefaultDebuggerPort),
+				localRegistry: "",
+			},
+		},
+		{
+			"Remote Local in config, no CLI",
+			config.GlobalConfig{
+				BuildkitHost:      "tcp://127.0.0.1:8372",
+				DebuggerHost:      "",
+				DebuggerPort:      config.DefaultDebuggerPort,
+				LocalRegistryHost: "",
+			},
+			noopArgs,
+			addresses{
+				buildkit:      "tcp://127.0.0.1:8372",
+				debugger:      fmt.Sprintf("tcp://127.0.0.1:%v", config.DefaultDebuggerPort),
+				localRegistry: "",
+			},
+		},
+		{
+			"Remote remote in config, no CLI",
+			config.GlobalConfig{
+				BuildkitHost:      "tcp://my-cool-host:8372",
+				DebuggerHost:      "",
+				DebuggerPort:      config.DefaultDebuggerPort,
+				LocalRegistryHost: "",
+			},
+			noopArgs,
+			addresses{
+				buildkit:      "tcp://my-cool-host:8372",
+				debugger:      fmt.Sprintf("tcp://my-cool-host:%v", config.DefaultDebuggerPort),
+				localRegistry: "",
+			},
+		},
+		{
+			"Nonstandard local in config, no CLI",
+			config.GlobalConfig{
+				BuildkitHost:      "docker-container://my-container",
+				DebuggerHost:      "",
+				DebuggerPort:      config.DefaultDebuggerPort,
+				LocalRegistryHost: "",
+			},
+			noopArgs,
+			addresses{
+				buildkit:      "docker-container://my-container",
+				debugger:      fmt.Sprintf("tcp://127.0.0.1:%v", config.DefaultDebuggerPort),
+				localRegistry: "",
+			},
+		},
+		{
+			"Debugger port specified",
+			config.GlobalConfig{
+				BuildkitHost:      "",
+				DebuggerHost:      "",
+				DebuggerPort:      5678,
+				LocalRegistryHost: "",
+			},
+			noopArgs,
+			addresses{
+				buildkit:      buildkitd.DockerAddress,
+				debugger:      "tcp://127.0.0.1:5678",
+				localRegistry: "",
+			},
+		},
+		{
+			"Debugger host and port specified",
+			config.GlobalConfig{
+				BuildkitHost:      "",
+				DebuggerHost:      "tcp://127.0.0.1:1234",
+				DebuggerPort:      5678,
```

Comment:
is this test supposed to pass when `1234` doesn't match `5678`?

Reply:
Yes. Specifying both means the host setting wins. The port setting is deprecated.

Reply:
More clarification:

The old code would only use port. If the feature flag was on, this setting would generate a deprecation notice. When unifying the codepaths, I made the preferably-deprecated setting fully deprecated, and the new setting always preferred. You can get old behavior by omitting the new setting and only specifying port.